### ISSUE TITLE
Change stroy content suggestions:

### ISSUE THREAD

**Issue by amrayach** [amrohendawi/unraveling-bert-article#3]
# TL;DR
This study investigates the limitations and scope of transfer learning on BERT (generic language model) and evaluates the factors and tradeoffs that influence the acquisition of a new downstream task. Since machine learning is on the rise in numerous problem-solving domains, particularly natural language processing, computationally superior methods are being introduced in the field of transfer learning to tackle that. In addition, transfer learning capitalizes on prior knowledge by repurposing it to solve and solve new problems. Since BERT learns linguistic patterns and correlations by processing large unlabeled text corpora, it is typically fine-tuned for new problems rather than retrained from scratch. However, fine-tuning a complex model such as BERT requires consideration of numerous underlying factors. We present some of the most relevant elements utilizing interactive visualization components and analyze them to provide the end-user with an overview that may aid them in future BERT fine-tuning endeavors. Furthermore, the results demonstrate that the system's effectiveness increases as the number of layers, fine-tuning epochs, and dataset size used for fine-tuning increase. Additionally, a BERT model with a shallower depth transfers better than one with a more significant depth. 

# BERT and Transfer-Learning
The Bidirectional Encoder Representations from Transformers (BERT) is a natural language processing (NLP) technique employed for training self-supervised learning language models. Moreover, BERT revolutionized how machines understand human language by following a bidirectional transformer design that learns deep bidirectional representations from an unlabeled corpus by jointly conditioning both left and right contexts in all layers. This learning process is achieved by using the simple tasks of Masked Language Modeling and Next Sentence Prediction. Finally, BERT generalized language understanding allows it to be extended to new domains such as text classification, question answering, and text generation.

Generally, many new tasks and domains have scars or low data resources, making the learning process harder from scratch. Moreover, the transfer learning technique was introduced to handle this issue, broadly transferring and leveraging already acquired knowledge onto new domains. Transfer learning is a learning procedure in which representations learned on a source task are transmitted to improve learning on the target task. In the case of machine learning, pre-trained models on source task A are repurposed by slightly fine-tuning the network's weights on a target task B. However, in the case of BERT, transfer learning tries to exploit its general language understanding to improve and achieve higher generalization in another one.

![diagram-20220808](https://user-images.githubusercontent.com/25460571/183492765-cf7df48c-7738-4ff7-a1f3-5e1e28367fde.png)

or 

![transfer](https://user-images.githubusercontent.com/25460571/183493069-b86d7aa5-2be1-4d46-90ea-4b3a2e873822.png)

Transfer learning formal definition:
Given a source domain $D_s$, a corresponding source task $T_s$, as well as a target domain $D_t$ and a target task $T_t$, the objective of transfer learning is to enable us to learn the target conditional probability distribution $P(Y_t|X_t)$ in $D_t$ with the information gained from $D_s$ and where $D_s \neq D_t$ or $T_s \neq T_t$

# Layer and epoch effect on transferability
A BERT model is fine-tuned on classification tasks with multiple labels to demonstrate layer and epoch effect on transferability by applying t-SNE on the layer's embeddings to reduce its dimensionality from 768 to 2 dimensions before and after fine-tuning to observe the changes on the layer level. As observed in different studies, the intermediate layers of BERT models tend to contain the most syntactic information, which is why these layers are the most transferable across tasks. Therefore, when using transfer learning with BERT models, it is essential to keep this in mind and focus on the middle layers. The higher layers seem more effective in this process because they acquire a generalized understanding. However, the lower layers seem to be more specialized and only understand specific aspects of the data. Finally, the number of epochs also seems to affect the performance of BERT models.

---

VISUALIZATION COMPONENT

---

The results of the following visualization hold two main observations:

 1. The data points of different classes are highly mixed, and no pattern or discrimination boundaries are yet developed at the beginning of the training loop. As the training progresses, an apparent clustering of the different classes starts to establish itself in some layers.
 2. The pattern and clustering of the different classes are primarily evident in the middle to higher layers of the model.
 3. The pre-trained Bert model has a low or non-existing understanding of unseen data, but after a proper fine-tuning procedure, it can generalize and adapt to new domains effectively.

**Comment by amrohendawi** [comment]
I appreciate the content you provided.
Most of the text hasn't changed from the current version, but I can see some modifications.
My next step will be to simplify it further for non-expert readers.
The two diagrams however are more suitable for a thesis or scientific report.
I will see how I can incorporate one of them in the final website version.

A sub-branch has been created for this issue.
